File: cpp/2187.cpp
#include <climits>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
  long long minimumTime(vector<int> &time, int totalTrips) {
    long long lb = 0, ub = LONG_LONG_MAX, ret;
    while (ub >= lb) {
      long long mid = lb + (ub - lb) / 2;
      unsigned long long cnt = 0;
      for (int i = 0; i < time.size(); i++)
        cnt += mid / static_cast<long long>(time[i]);
      if (cnt >= totalTrips)
        ret = mid, ub = mid - 1;
      else
        lb = mid + 1;
    }
    return ret;
  }
};

File: go/2187.go
package main

import "math"

func minimumTime(time []int, totalTrips int) int64 {
	var ret, l, r int64 = 0, 0, math.MaxInt64
	for l <= r {
		var cnt int64 = 0
		m := (l + r) >> 1
		for _, v := range time {
			cnt += m / int64(v)
			if cnt > int64(totalTrips) {
				break
			}
		}
		if cnt >= int64(totalTrips) {
			ret = m
			r = m - 1
		} else {
			l = m + 1
		}
	}
	return ret
}
